Implement RBAC changes

prabushyam opened the pull request:
@mcieplak

mcieplak commented on pkg/istio/servicerole/servicerole.go:
```diff
@@ -53,14 +75,151 @@ func (srMgr *ServiceRoleMgr) GetServiceRoleMap() (map[string]*ServiceRoleInfo, e
+			continue
+		}
+
+		if assertionRole != roleName {
```
Shouldn't the assertions already be grouped by the role at this point? Why are we doing this check?

prabushyam replied:
yep, this is extra safety check, since the actual grouping was done by athenz.model, and this one is rbac provider implementation specific, wanted to have the check explicit

mcieplak commented on pkg/istio/servicerole/servicerole.go:
```diff
@@ -53,14 +75,151 @@ func (srMgr *ServiceRoleMgr) GetServiceRoleMap() (map[string]*ServiceRoleInfo, e
+			continue
+		}
+
+		rule := &v1alpha1.AccessRule{}
```
Can we create one object here and set all these values?

mcieplak commented on pkg/istio/servicerole/servicerole.go:
```diff
@@ -53,14 +75,151 @@ func (srMgr *ServiceRoleMgr) GetServiceRoleMap() (map[string]*ServiceRoleInfo, e
+		rule.Constraints = []*v1alpha1.AccessRule_Constraint{constraint}
+		rule.Methods = []string{method}
+		if path != "" {
+			rule.Paths = []string{path}
```
Will this allow all paths by default if not specified?

prabushyam replied:
yup, that should be expected behavior

mcieplak commented on pkg/controller/controller.go:
```diff
@@ -89,6 +91,25 @@ func (c *Controller) sync() error {
+		domainRBAC := m.ConvertAthenzPoliciesInToRBACModel(signedDomain.Domain)
+
+		provider := rbacv1.NewProvider()
+		rbacCRs := provider.ConvertAthenzModelIntoIstioRBAC(domainRBAC)
```
I am thinking we need to add the validation functions calls somewhere for the specs.

prabushyam replied:
good catch. totally missed the validation. added after generating each SR/SRB spec...

mcieplak commented on pkg/istio/servicerole/servicerole_test.go:
```diff
@@ -48,3 +50,248 @@ func TestCreateServiceRole(t *testing.T) {
+						Paths: []string{
+							"/protected/path",
+						},
+						Constraints: []*v1alpha1.AccessRule_Constraint{
```
Can we add a test case here without a path?

mcieplak commented on pkg/istio/servicerolebinding/servicerolebinding.go:
```diff
@@ -105,6 +108,75 @@ func (srbMgr *ServiceRoleBindingMgr) getSubjects(members []zms.MemberName) []*v1
+	subjects := make([]*v1alpha1.Subject, 0)
+	for _, member := range members {
+
+		//TODO: handle member.Expiration for expired members, for now ignore expiration
```
Does Athenz themselves remove the expired members?

prabushyam replied:
I think ZTS / ZMS does the expiration check when the role members are accessed (during checkAccess call), so ideally we should be doing while we get members of a role during every sync

mcieplak commented on pkg/istio/servicerolebinding/servicerolebinding.go:
```diff
@@ -19,7 +21,8 @@ import (
 const (
-	allUsers = "user.*"
+	allUsers         = "user.*"
+	wildCardAllUsers = "*"
```
Does istio have a constant for this wildcard use case which we could use?

prabushyam replied:
Couldn't find exportable constant for wildcard, let me know if you found one

mcieplak commented on pkg/istio/servicerolebinding/servicerolebinding_test.go:
```diff
@@ -40,3 +43,111 @@ func TestCreateServiceRoleBinding(t *testing.T) {
+			},
+			expectedMember: "*",
+			expectedErr:    nil,
+		},
```
Can we also add one test case with an invalid member name.

mcieplak commented on pkg/istio/servicerolebinding/servicerolebinding_test.go:
```diff
@@ -40,3 +43,111 @@ func TestCreateServiceRoleBinding(t *testing.T) {
+				members:  nil,
+			},
+			expectedSpec: nil,
+			expectedErr:  fmt.Errorf("no subjects found for the ServiceRoleBinding: "),
```
shouldn't this have the role name at the end of the string?

prabushyam replied:
the passed-in role name is empty, so the error message has nothing to print

mcieplak commented on pkg/istio/servicerolebinding/servicerolebinding_test.go:
```diff
@@ -40,3 +43,111 @@ func TestCreateServiceRoleBinding(t *testing.T) {
+					},
+				},
+			},
+			expectedErr: nil,
```
Can we also add an invalid member test case?

mcieplak commented:
@prabushyam Mostly looks good overall, had mainly minor comments. Unfortunately you will have to resolve the merge conflicts as well.

prabushyam commented on pkg/istio/servicerole/servicerole.go:
```diff
@@ -53,14 +75,151 @@ func (srMgr *ServiceRoleMgr) GetServiceRoleMap() (map[string]*ServiceRoleInfo, e
 }
 
+// NewConfig returns a new model.Config resource with the given namespace/name and spec
+func NewConfig(namespace string, name string, spec *v1alpha1.ServiceRole) model.Config {
```
will remove the createServiceRole function with a subsequent PR

prabushyam commented on pkg/istio/servicerolebinding/servicerolebinding.go:
```diff
@@ -105,6 +108,75 @@ func (srbMgr *ServiceRoleBindingMgr) getSubjects(members []zms.MemberName) []*v1
 }
 
+// NewConfig returns a new model.Config resource with the given namespace/name and spec
+func NewConfig(namespace string, name string, spec *v1alpha1.ServiceRoleBinding) model.Config {
```
sure, moved to common/util.go

prabushyam commented on pkg/istio/rbac/common/util.go:
```diff
@@ -0,0 +1,35 @@
+// e.g. app-domain:role.reader -> reader
+func ParseRoleFQDN(domainName zms.DomainName, roleFQDN string) (string, error) {
+	roleName := strings.TrimPrefix(roleFQDN, fmt.Sprintf("%s:role.", domainName))
+	if strings.Contains(roleName, ":") {
```
yep, trust roles can have another domain prefixed role names, which at the moment we cannot support. We need to think through that use case.

prabushyam commented on pkg/istio/rbac/common/util.go:
```diff
@@ -0,0 +1,35 @@
+
+// PrincipalToSPIFFE converts the Athenz principal into a SPIFFE compliant format
+// e.g. client-domain.frontend.some-app -> client-domain.frontend/sa/some-app
+func PrincipalToSPIFFE(principal string) (string, error) {
```
For handling the `<domain>:role.` prefix extraction? I couldn't find one in Go...

prabushyam commented on pkg/istio/rbac/common/util_test.go:
```diff
@@ -0,0 +1,93 @@
+
+	for _, c := range cases {
+		gotRole, gotErr := ParseRoleFQDN(c.domainName, c.roleFQDN)
+		a := assert.New(t)
```
assert.Equal() takes in `t` , I guess the current approach would re-user the `assert.New(t)` object for all the assertions

prabushyam commented:
@mcieplak  thank you for the review. I marked the fixed comments as resolved, and left the ones unresolved that I responded in-line. Please re-review when you get a chance

prabushyam commented:
I added the validation and that caught 2 issues:
- ServiceRole needs to have at least 1 `service` in the rule in 1.1 release, this validation check was removed in 1.2. So, I added `services: *` for all rules automatically. 
- ServiceRoleBinding RoleRef validation looks for `ServiceRole` string instead of `service-role` that the current code was trying to do, fixed both.

mcieplak commented on pkg/athenz/model_test.go:
```diff
@@ -0,0 +1,431 @@
+	"github.com/ardielle/ardielle-go/rdl"
+	"github.com/yahoo/athenz/clients/go/zms"
+	"reflect"
+	"testing"
```
Can we update the imports in main.go and model_test.go as well?
